### PR TITLE
mtmd: Add DeepSeekOCR 2 Support

### PR DESCRIPTION
**Pull request by sfallah**
## Overview

This PR adds support for DeepSeek-OCR-2 [deepseek-ai/DeepSeek-OCR-2](https://huggingface.co/deepseek-ai/DeepSeek-OCR-2).

GGUF model files for testing this PR are available at [sabafallah/DeepSeek-OCR-2-GGUF](https://huggingface.co/sabafallah/DeepSeek-OCR-2-GGUF).

The implementation adds the DeepSeek-OCR-2 vision path that reuses SAM encoder (from DeepSeek-OCR v1), a new Qwen2-based vision encoder, and multi-tile dynamic-resolution image preprocessing.

The preprocessing includes multi-tile dynamic-resolution, with a 1024 global view and grid of tiles with 768 tile views.
The image resize/padding behavior is Pillow-based to match the original deepseek-ai implementation.

### Main changes

- added DeepSeek-OCR-2 model support in mtmd
- reused the SAM implementation from DeepSeek-OCR where possible
- added the Qwen2 vision encoder
- added multi-tile dynamic-resolution preprocessing
- implemented the DeepSeek-OCR-2 attention mask handling for the Qwen2 encoder
- updated `mtmd_encode` to handle multi-tile and global view token counts diff
- added converter support for DeepSeek-OCR-2 vision weights
- extended `tools/mtmd/tests/test-deepseek-ocr.py` to cover both DeepSeek-OCR and DeepSeek-OCR-2

### Implementation notes

- The SAM encoder is shared with DeepSeek-OCR through the existing `build_sam`.
- The Qwen2 attention mask is prepared CPU-side.
- The dynamic-resolution preprocessing follows the InternVL-style behavior used by the reference implementation.

### Testing

- Extended `tools/mtmd/tests/test-deepseek-ocr.py` to cover both v1 and v2.
- DeepSeek-OCR-2 result: CER 0.6944 / chrF 23.01
- DeepSeek-OCR-2 HF reference: CER 0.7761 / chrF 28.70
- Manually verified with `llama-mtmd-cli`.
- Manually verified with `llama-server`.

### Caveats

- The DeepSeek-OCR-2 test gate is intentionally loose (`cer_tol = 0.12`). The current test image is low quality, and the HF reference itself has a high CER on this sample.
- The DRY sampler is used as an approximation for the HF `no_repeat_ngram_size` behavior.
- For DeepSeek-OCR-2, `llama-server` currently requires `--chat-template deepseek-ocr --no-jinja`.  This is due to the template handling for DeepSeek-OCR-2.

- Running `llama-server`:
  ```bash
  build/bin/llama-server \
  -m gguf_models/deepseek-ai/deepseek-ocr-2-bf16.gguf \
  --mmproj gguf_models/deepseek-ai/mmproj-deepseek-ocr-2-bf16.gguf \
  --chat-template deepseek-ocr --no-jinja \
  --temp 0 \
  --flash-attn off \
  --no-warmup \
  -n 2048 \
  --dry-multiplier 0.8 --dry-base 1.75 --dry-allowed-length 2 \
  --dry-penalty-last-n -1 --dry-sequence-breaker none
  ```

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - I used AI assistance for code review, debugging, implementation checks, and testing. I have reviewed the submitted changes and take responsibility for the full contents of this PR.

### CODE REVIEW

File: tools/mtmd/models/deepseekocr2.cpp
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp:1]: IIRC, the vision encoder is mostly the same between DS-OCR v1 and v2, it's better to find a way to reuse the code from v1 when the other PR is merged

Edit: even better, v2 vision encoder is the stripped down version of v1, so it should be trivial to merge these 2 into one file

**Comment by ngxson** [comment]
you may also need to enable `mtmd_decode_use_non_causal` for this model, so that text is causal while image tokens are non-causal

**Comment by ngxson** [comment]
hmm, please ignore what I said earlier, it seems like their paper is poorly written and the v2 is not easier than v1.

image

basically it's a 3-step pipeline: SAM encode the image into embeddings, Qwen take embedding in and generate yet another embeddings, then finally they are fed into the main deepseek model

~~if my understanding is correct, that means the prompt will now be processed by vision encoder, which make things a whole lot more complicated~~

**Comment by ngxson** [comment]
small correction, seems like what they refers to "query" on the diagram is not the text prompts, but simply a set of pre-trained tokens (fixed size), so that won't be too complicated to implement

the only thing that need to be provided correctly to the cgraph is the attention mask that will correctly mask the image tokens as non-causal and the query as causal, it should not be complicated

and finally, because SAM is the same between 2 models it is recommended that you extract SAM as a function and inherit it in DS-OCR v2, example:

```cpp
struct clip_graph_deepseekocr : clip_graph {
    clip_graph_deepseekocr(clip_ctx * ctx, const clip_image_f32 & img) : clip_graph(ctx, img) {}
    ggml_cgraph * build() override;
    ggml_cgraph * build_sam(); // build the SAM model
};

struct clip_graph_deepseekocr2 : clip_graph_deepseekocr { // inherit
    clip_graph_deepseekocr2(clip_ctx * ctx, const clip_image_f32 & img) : clip_graph(ctx, img) {}
    ggml_cgraph * build() override; // can directly reuse build_sam() from base class, no need to duplicate the code
};
```

File: tools/mtmd/models/deepseekocr2.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        ggml_tensor * inpL = inp;
+
+        ggml_tensor * inp_pos = ggml_cast(ctx0, ggml_arange(ctx0, 0, seq_len, 1), GGML_TYPE_I32);
+
+        int  num_heads = 14;
+        auto head_dim  = n_embd / num_heads;
+        int  kv_heads  = 2;
+
+        // qwen2 layers
+        for (int il = 0; il < n_layer; ++il) {
```
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp]: this loop should be able to be replaced by build_vit, it should already support attention mask

File: conversion/base.py [hunk 2]
```diff
@@ -1114,7 +1114,7 @@ def filter_tensors(cls, item: tuple[str, Callable[[], Tensor]]) -> tuple[str, Ca
         # Skip multimodal tensors
         if name.startswith(("mlp", "vit.", "vpm.", "siglip2.", "conformer.", "merger.", "resampler.", "sound_encoder.", "sound_projection.", "speech_embeddings.")) \
                 or "visual." in name or "vision." in name or "audio." in name or "talker." in name \
-                or "vision_" in name or "audio_" in name or "sam_model" in name \
+                or "vision_" in name or "audio_" in name or "sam_model" in name or "qwen2_model" in name \
```
Comment (CISC) [inline conversion/base.py]: I'm not comfortable with this in `TextModel`, please move it to `DeepseekModel`.

File: tools/mtmd/models/deepseekocr2.cpp (hunk 1 above)
Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp]: `build_vit` does not support GQA right now.
I prepared a small PR for this here: https://github.com/ggml-org/llama.cpp/pull/23782
The change is basically to use `n_head_kv` for the K/V reshape, and fall back to `n_head` when it is not set.

But to be honest, I am a bit conflicted about this. 
I am not comfortable changing `build_vit`.
Please give me some guidance here.

Comment (ngxson) [reply tools/mtmd/models/deepseekocr2.cpp]: if GQA is the only thing missing from build_vit, then it's ok to add. there are ViT models use GQA, so it make sense to support that

except from GQA, is anything else missing?

Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp]: no,
the tests in `mtmd/tests.sh `are running right now, if successful I will open https://github.com/ggml-org/llama.cpp/pull/23782

File: conversion/base.py (hunk 2 above)
Comment (sfallah) [reply conversion/base.py]: Moved qwen2_model (and sam_model, while at it) out of TextModel.

File: tools/mtmd/models/deepseekocr2.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    ggml_tensor * cur;
+
+    cur = ggml_mul_mat(ctx0, model.mm_fc_w, qwen2_out);
+    cur = ggml_add(ctx0, cur, model.mm_fc_b);
+
+    const auto n_dim = cur->ne[0];
+
+    // separator row, only after the global view (256 tokens)
+    if (cur->ne[1] == 256) {
```
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp]: this pattern is not very clean.

instead, it's better to add a new field to `clip_image_f32`, let's say `bool clip_image_f32::add_viewsep`

the preprocessor can set this field on the f32 image before returning to the caller, and here we check for `img->add_viewsep`

File: tools/mtmd/clip.cpp [hunk 4]
```diff
@@ -3293,6 +3305,15 @@ int clip_n_output_tokens(const struct clip_ctx * ctx, struct clip_image_f32 * im
                 int oh = (img->ny / patch_size) / merge;
                 n_patches = (ow + 1) * oh + 2;
             } break;
+        case PROJECTOR_TYPE_DEEPSEEKOCR2:
+        {
+            // 1024 global view -> 256 query tokens + 1 view separator = 257;
+            // 768 local tile   -> 144 query tokens, no separator.
+            n_patches /= 16;
+            if (n_patches == 256) {
```
Comment (ngxson) [inline tools/mtmd/clip.cpp]: same here, better to check for image->add_viewsep

File: conversion/deepseek.py [hunk 5]
```diff
@@ -49,20 +53,20 @@ def get_vision_config(self) -> dict[str, Any]:
             raise ValueError("DeepseekOCR model requires 'vision_config' in the model configuration, but it was not found")
 
         vision_config['sam'] = vision_config['width']['sam_vit_b']
-        vision_config.update(vision_config['width']['clip-l-14-224'])
-        vision_config['hidden_size'] = vision_config['width']
-        vision_config['num_heads'] = vision_config['heads']
-        vision_config['intermediate_size'] = vision_config['heads'] * 4
+        if vision_config['width'].get('clip-l-14-224') is not None:
+            vision_config.update(vision_config['width']['clip-l-14-224'])
+        if isinstance(vision_config['width'], int):
```
Comment (CISC) [inline conversion/deepseek.py:58]: This will never be true because you will hit an `AttributeError` above.

Edit: Or is it that `clip-l-14-224` contains `width`?

Comment (sfallah) [reply conversion/deepseek.py:58]: yes, 'clip-l-14-224' has also a 'width'

File: tools/mtmd/models/deepseekocr2.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        inp = ggml_cont(ctx0, ggml_permute(ctx0, inp, 1, 0, 2, 3));
+
+        auto num_image_tokens = inp->ne[1]; // H*W
+
+        // query based on numbers of image tokens (in SAM output)
+        // 16x16 -> query_1024 (1024x1024 images)
+        // 12x12 -> query_768 (768x768 images)
+
+        ggml_tensor * query_embed = model.resample_query_1024;
+        int           num_queries = 256;
```
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp:28]: maybe have a GGML_ASSERT to make sure the `num_image_tokens` can be either 144 or 256

File: tools/mtmd/models/deepseekocr2.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    ggml_tensor * cur;
+
+    cur = ggml_mul_mat(ctx0, model.mm_fc_w, qwen2_out);
+    cur = ggml_add(ctx0, cur, model.mm_fc_b);
+
+    const auto n_dim = cur->ne[0];
+
+    // separator row, only after the global view
+    if (img.add_viewsep) {
+        ggml_tensor * vs = ggml_reshape_2d(ctx0, model.view_seperator, n_dim, 1); // (n_dim, 1)
```
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp]: this reshape should be done at model conversion

File: tools/mtmd/models/deepseekocr2.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    ggml_tensor * inp_raw = build_inp_raw();
+
+    ggml_tensor * sam_out = build_sam(inp_raw);
+
+    ggml_tensor * qwen2_out;
+    // Building Qwen2 encoder
+    {
+        ggml_tensor * inp;
+
+        inp = ggml_cpy(ctx0, sam_out, ggml_dup_tensor(ctx0, sam_out));
```
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp]: I don't quite get this, why we need to copy the tensor?

Comment (ngxson) [reply tools/mtmd/models/deepseekocr2.cpp]: same question for `clip_graph_deepseekocr`

File: tools/mtmd/models/deepseekocr2.cpp (hunk 6 above)
Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp:28]: Added

File: tools/mtmd/models/deepseekocr2.cpp (hunk 7 above)
Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp]: done

File: tools/mtmd/models/deepseekocr2.cpp (hunk 8 above)
Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp]: These were leftovers from debugging.
I was dumping sam_out and wanted to isolate it before possible mutation effects, to be on the safe side.
Redundant ggml_cpy ops for sam_out and others are removed now.

File: tools/mtmd/models/deepseekocr2.cpp [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    ggml_tensor * qwen2_out;
+    // Building Qwen2 encoder
+    {
+        ggml_tensor * inp;
+
+        inp = ggml_reshape_2d(ctx0, sam_out, sam_out->ne[0] * sam_out->ne[1], sam_out->ne[2]); // H*W, C
+        inp = ggml_cont(ctx0, ggml_permute(ctx0, inp, 1, 0, 2, 3));
+
+        auto num_image_tokens = inp->ne[1]; // H*W
+        GGML_ASSERT(num_image_tokens == 144 ^ num_image_tokens == 256);
```
Comment (ngxson) [inline tools/mtmd/models/deepseekocr2.cpp]:
```suggestion
        GGML_ASSERT(num_image_tokens == 144 || num_image_tokens == 256);
```

**Review by ngxson — APPROVED** [review]

File: tools/mtmd/models/deepseekocr2.cpp (hunk 9 above)
Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp]: It practically makes no difference if it is XOR or OR.
XOR is logically more accurate.
Because it is either a tile image or a global view image.
But we can go with `||`.

Comment (ngxson) [reply tools/mtmd/models/deepseekocr2.cpp]: I don't see how XOR is more accurate. are you saying to me that `num_image_tokens` can be both 144 AND 256 at the same time, how?

And even that, developers never write this in practice because it's dangerous. The whole code base of llama.cpp never have such pattern.

For example, say a=1 and b=2, a XOR b = 3 but !!a XOR !!b == false. Simply write using && || is much safer and easier to understand. So think twice before adding something new to the coding convention.

Comment (sfallah) [reply tools/mtmd/models/deepseekocr2.cpp]: yes but we didn't had `a ^ b`
we have `a==1 ^ b==2`
anyway I see your point
my first commit that i than squashed/force-pushed was exactly what you suggested.

**Review by CISC — APPROVED** [review]

**Comment by ngxson** [comment]
@sfallah can you check locally if `test-llama-archs` passes? the CI fails on that test but I'm not sure if it has something to do with this PR, since the test only runs the text model

**Comment by sfallah** [comment]
> @sfallah can you check locally if `test-llama-archs` passes? the CI fails on that test but I'm not sure if it has something to do with this PR, since the test only runs the text model

@ngxson

All tests pass locally.
As far as I can tell, this test is also failing on `master`.
`DSOCR-2` is currently excluded from `test-llama-archs`. In fact, I have already prepared a small follow-up PR that adds DeepSeek-OCR, since you have already approved this PR.
